NullPointer exception while registering custom data table converter

## Summary

NullPointer exception while registering custom data table converter

## Current Behavior
Register the following table converter:
```java
    @DataTableType
    public List convertDataTable(DataTable dataTable) {
        return new ArrayList();
    }
``` 
Define the following step:
```java
    @Given("table {}=")
    public void setCustomDataTable(String name, List value) {
        log.debug("Param {} = {}", name, value);
    }
```
Create the following scenario:
```gherkin
Scenario: Compare empty data tables
    Given table empty_table=
      | test |
      |      |
```
Run the scenario:
```
Step failed
io.cucumber.core.exception.CucumberException: Could not convert arguments for step [table {}=] defined at 'com.cucumber.utils.context.stepdefs.ParamSteps.setCustomDataTable(java.lang.String,java.util.Map)'.
The details are in the stacktrace below.
	at io.cucumber.core.runner.PickleStepDefinitionMatch.couldNotConvertArguments(PickleStepDefinitionMatch.java:108)
	at io.cucumber.core.runner.PickleStepDefinitionMatch.runStep(PickleStepDefinitionMatch.java:52)
	at io.cucumber.core.runner.TestStep.executeStep(TestStep.java:64)
	at io.cucumber.core.runner.TestStep.run(TestStep.java:49)
	at io.cucumber.core.runner.PickleStepTestStep.run(PickleStepTestStep.java:46)
	at io.cucumber.core.runner.TestCase.run(TestCase.java:51)
	at io.cucumber.core.runner.Runner.runPickle(Runner.java:66)
	at io.cucumber.core.runtime.Runtime.lambda$null$2(Runtime.java:100)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at io.cucumber.core.runtime.Runtime$SameThreadExecutorService.execute(Runtime.java:243)
	at java.base/java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:118)
	at io.cucumber.core.runtime.Runtime.lambda$run$3(Runtime.java:100)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.SliceOps$1$1.accept(SliceOps.java:199)
	at java.base/java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1631)
	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:127)
	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:502)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:488)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at io.cucumber.core.runtime.Runtime.run(Runtime.java:101)
	at io.cucumber.core.cli.Main.run(Main.java:73)
	at io.cucumber.core.cli.Main.main(Main.java:31)
Caused by: io.cucumber.datatable.CucumberDataTableException: 'java.util.List' could not transform
      | test |
      |      |

	at io.cucumber.datatable.DataTableType.transform(DataTableType.java:124)
	at io.cucumber.datatable.DataTableTypeRegistryTableConverter.toListOrNull(DataTableTypeRegistryTableConverter.java:150)
	at io.cucumber.datatable.DataTableTypeRegistryTableConverter.convertEntryKeys(DataTableTypeRegistryTableConverter.java:262)
	at io.cucumber.datatable.DataTableTypeRegistryTableConverter.toMap(DataTableTypeRegistryTableConverter.java:216)
	at io.cucumber.datatable.DataTableTypeRegistryTableConverter.convert(DataTableTypeRegistryTableConverter.java:73)
	at io.cucumber.datatable.DataTable.convert(DataTable.java:357)
	at io.cucumber.core.stepexpression.StepExpressionFactory.lambda$createExpression$4(StepExpressionFactory.java:63)
	at io.cucumber.core.stepexpression.DataTableArgument.getValue(DataTableArgument.java:19)
	at io.cucumber.core.runner.PickleStepDefinitionMatch.runStep(PickleStepDefinitionMatch.java:47)
	... 24 more
Caused by: java.lang.NullPointerException
	at io.cucumber.java.AbstractDatatableElementTransformerDefinition.replaceEmptyPatternsWithEmptyString
```

Heya, 

I can't reproduce this using the information you've provided and the stack trace got cut off a the critical part

```
Caused by: java.lang.NullPointerException
	at io.cucumber.java.AbstractDatatableElementTransformerDefinition.replaceEmptyPatternsWithEmptyString
```

There should be some line numbers here and perhaps a few more bits of stack frame.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -19,7 +19,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 ### Removed
 
 ### Fixed
-
+ * [Java] Fix NPE in AbstractDatatableElementTransformerDefinition ([#1865](https://github.com/cucumber/cucumber-jvm/pull/1865) Florin Slevoaca, M.P. Korstanje, ) 
+ 
 ## [5.0.0] (2020-01-16) - [Release Notes](release-notes/v5.0.0.md)
 
 ### Added

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -20,7 +20,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/).
 ### Removed
 
 ### Fixed
- * [Java] Fix NPE in AbstractDatatableElementTransformerDefinition ([#1865](https://github.com/cucumber/cucumber-jvm/pull/1865) Florin Slevoaca, M.P. Korstanje, ) 
+ * [Java/Java8] Fix NPE in AbstractDatatableElementTransformerDefinition ([#1865](https://github.com/cucumber/cucumber-jvm/pull/1865) Florin Slevoaca, M.P. Korstanje)
+ * [Core] Fix collision when using `Datatable.asMaps` ([cucumber/cucumber#877](https://github.com/cucumber/cucumber/pull/877) M.P. Korstanje)
  
 ## [5.0.0] (2020-01-16) - [Release Notes](release-notes/v5.0.0.md)
 

--- a/java/src/main/java/io/cucumber/java/AbstractDatatableElementTransformerDefinition.java
+++ b/java/src/main/java/io/cucumber/java/AbstractDatatableElementTransformerDefinition.java
@@ -5,12 +5,12 @@ import io.cucumber.datatable.DataTable;
 
 import java.lang.reflect.Method;
 import java.util.ArrayList;
+import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
 
 import static io.cucumber.datatable.DataTable.create;
 import static java.util.stream.Collectors.toList;
-import static java.util.stream.Collectors.toMap;
 
 public class AbstractDatatableElementTransformerDefinition extends AbstractGlueDefinition {
     private final String[] emptyPatterns;
@@ -36,14 +36,19 @@ public class AbstractDatatableElementTransformerDefinition extends AbstractGlueD
     }
 
     Map<String, String> replaceEmptyPatternsWithEmptyString(Map<String, String> fromValue) {
-        return fromValue.entrySet().stream()
-            .collect(toMap(
-                entry -> replaceEmptyPatternsWithEmptyString(entry.getKey()),
-                entry -> replaceEmptyPatternsWithEmptyString(entry.getValue()),
-                (s, s2) -> {
-                    throw createDuplicateKeyAfterReplacement(fromValue);
-                }
-            ));
+        Map<String, String> replacement = new LinkedHashMap<>();
+
+        fromValue.forEach((String key, String value) -> {
+            String potentiallyEmptyKey = replaceEmptyPatternsWithEmptyString(key);
+            String potentiallyEmptyValue = replaceEmptyPatternsWithEmptyString(value);
+
+            if (replacement.containsKey(potentiallyEmptyKey)) {
+                throw createDuplicateKeyAfterReplacement(fromValue);
+            }
+            replacement.put(potentiallyEmptyKey, potentiallyEmptyValue);
+        });
+
+        return replacement;
     }
 
     private IllegalArgumentException createDuplicateKeyAfterReplacement(Map<String, String> fromValue) {
@@ -59,7 +64,7 @@ public class AbstractDatatableElementTransformerDefinition extends AbstractGlueD
 
     String replaceEmptyPatternsWithEmptyString(String t) {
         for (String emptyPattern : emptyPatterns) {
-            if (t.equals(emptyPattern)) {
+            if (emptyPattern.equals(t)) {
                 return "";
             }
         }

--- a/java/src/main/java/io/cucumber/java/AbstractDatatableElementTransformerDefinition.java
+++ b/java/src/main/java/io/cucumber/java/AbstractDatatableElementTransformerDefinition.java
@@ -12,7 +12,7 @@ import java.util.Map;
 import static io.cucumber.datatable.DataTable.create;
 import static java.util.stream.Collectors.toList;
 
-public class AbstractDatatableElementTransformerDefinition extends AbstractGlueDefinition {
+class AbstractDatatableElementTransformerDefinition extends AbstractGlueDefinition {
     private final String[] emptyPatterns;
 
     AbstractDatatableElementTransformerDefinition(Method method, Lookup lookup, String[] emptyPatterns) {

--- a/java/src/test/java/io/cucumber/java/JavaDataTableTypeDefinitionTest.java
+++ b/java/src/test/java/io/cucumber/java/JavaDataTableTypeDefinitionTest.java
@@ -84,7 +84,7 @@ class JavaDataTableTypeDefinitionTest {
         Method method = JavaDataTableTypeDefinitionTest.class.getMethod("converts_table_entry_to_string", Map.class);
         JavaDataTableTypeDefinition definition = new JavaDataTableTypeDefinition(method, lookup, new String[]{"[empty]"});
         assertThat(definition.dataTableType().transform(emptyTable.asLists()),
-            is(singletonList("converts_table_entry_to_string={=d, a=}")));
+            is(singletonList("converts_table_entry_to_string={a=, =d}")));
     }
 
     public String converts_table_entry_to_string(Map<String, String> entry) {

--- a/java/src/test/java/io/cucumber/java/JavaDefaultDataTableCellTransformerDefinitionTest.java
+++ b/java/src/test/java/io/cucumber/java/JavaDefaultDataTableCellTransformerDefinitionTest.java
@@ -39,6 +39,14 @@ class JavaDefaultDataTableCellTransformerDefinitionTest {
         assertThat(transformed, is("transform_string_to_type="));
     }
 
+    @Test
+    void can_transform_null_while_using_replacement_patterns() throws Throwable {
+        Method method = JavaDefaultDataTableCellTransformerDefinitionTest.class.getMethod("transform_string_to_type", String.class, Type.class);
+        JavaDefaultDataTableCellTransformerDefinition definition = new JavaDefaultDataTableCellTransformerDefinition(method, lookup, new String[]{"[empty]"});
+        Object transformed = definition.tableCellByTypeTransformer().transform(null, String.class);
+        assertThat(transformed, is("transform_string_to_type=null"));
+    }
+
     public Object transform_string_to_type(String fromValue, Type toValueType) {
         return "transform_string_to_type=" + fromValue;
     }

--- a/java/src/test/java/io/cucumber/java/JavaDefaultDataTableEntryTransformerDefinitionTest.java
+++ b/java/src/test/java/io/cucumber/java/JavaDefaultDataTableEntryTransformerDefinitionTest.java
@@ -78,7 +78,7 @@ class JavaDefaultDataTableEntryTransformerDefinitionTest {
             () -> definition.tableEntryByTypeTransformer().transform(fromValue, String.class, cellTransformer)
         );
 
-        assertThat(exception.getMessage(), is("After replacing [empty] and [blank] with empty strings the datatable entry contains duplicate keys: {[blank]=b, [empty]=a}"));
+        assertThat(exception.getMessage(), is("After replacing [empty] and [blank] with empty strings the datatable entry contains duplicate keys: {[empty]=a, [blank]=b}"));
     }
 
     public <T> T correct_method(Map<String, String> fromValue, Type toValueType) {

--- a/java/src/test/java/io/cucumber/java/JavaDefaultDataTableEntryTransformerDefinitionTest.java
+++ b/java/src/test/java/io/cucumber/java/JavaDefaultDataTableEntryTransformerDefinitionTest.java
@@ -6,7 +6,7 @@ import org.junit.jupiter.api.Test;
 
 import java.lang.reflect.Method;
 import java.lang.reflect.Type;
-import java.util.HashMap;
+import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.stream.Collectors;
@@ -54,8 +54,19 @@ class JavaDefaultDataTableEntryTransformerDefinitionTest {
     }
 
     @Test
+    void transforms_nulls_with_correct_method() throws Throwable {
+        Map<String, String> fromValue = singletonMap("key", null);
+        Method method = JavaDefaultDataTableEntryTransformerDefinitionTest.class.getMethod("correct_method", Map.class, Type.class);
+        JavaDefaultDataTableEntryTransformerDefinition definition =
+            new JavaDefaultDataTableEntryTransformerDefinition(method, lookup, false, new String[]{"[empty]"});
+
+        assertThat(definition.tableEntryByTypeTransformer()
+            .transform(fromValue, String.class, cellTransformer), is("key=null"));
+    }
+
+    @Test
     void throws_for_multiple_empties_with_correct_method() throws Throwable {
-        Map<String, String> fromValue = new HashMap<>();
+        Map<String, String> fromValue = new LinkedHashMap<>();
         fromValue.put("[empty]", "a");
         fromValue.put("[blank]", "b");
         Method method = JavaDefaultDataTableEntryTransformerDefinitionTest.class.getMethod("correct_method", Map.class, Type.class);

--- a/java8/src/main/java/io/cucumber/java8/AbstractDatatableElementTransformerDefinition.java
+++ b/java8/src/main/java/io/cucumber/java8/AbstractDatatableElementTransformerDefinition.java
@@ -3,12 +3,12 @@ package io.cucumber.java8;
 import io.cucumber.datatable.DataTable;
 
 import java.util.ArrayList;
+import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
 
 import static io.cucumber.datatable.DataTable.create;
 import static java.util.stream.Collectors.toList;
-import static java.util.stream.Collectors.toMap;
 
 public class AbstractDatatableElementTransformerDefinition extends AbstractGlueDefinition {
     private final String[] emptyPatterns;
@@ -34,14 +34,19 @@ public class AbstractDatatableElementTransformerDefinition extends AbstractGlueD
     }
 
     Map<String, String> replaceEmptyPatternsWithEmptyString(Map<String, String> fromValue) {
-        return fromValue.entrySet().stream()
-            .collect(toMap(
-                entry -> replaceEmptyPatternsWithEmptyString(entry.getKey()),
-                entry -> replaceEmptyPatternsWithEmptyString(entry.getValue()),
-                (s, s2) -> {
-                    throw createDuplicateKeyAfterReplacement(fromValue);
-                }
-            ));
+        Map<String, String> replacement = new LinkedHashMap<>();
+
+        fromValue.forEach((String key, String value) -> {
+            String potentiallyEmptyKey = replaceEmptyPatternsWithEmptyString(key);
+            String potentiallyEmptyValue = replaceEmptyPatternsWithEmptyString(value);
+
+            if (replacement.containsKey(potentiallyEmptyKey)) {
+                throw createDuplicateKeyAfterReplacement(fromValue);
+            }
+            replacement.put(potentiallyEmptyKey, potentiallyEmptyValue);
+        });
+
+        return replacement;
     }
 
     private IllegalArgumentException createDuplicateKeyAfterReplacement(Map<String, String> fromValue) {
@@ -57,7 +62,7 @@ public class AbstractDatatableElementTransformerDefinition extends AbstractGlueD
 
     String replaceEmptyPatternsWithEmptyString(String t) {
         for (String emptyPattern : emptyPatterns) {
-            if (t.equals(emptyPattern)) {
+            if (emptyPattern.equals(t)) {
                 return "";
             }
         }

--- a/java8/src/main/java/io/cucumber/java8/AbstractDatatableElementTransformerDefinition.java
+++ b/java8/src/main/java/io/cucumber/java8/AbstractDatatableElementTransformerDefinition.java
@@ -10,7 +10,7 @@ import java.util.Map;
 import static io.cucumber.datatable.DataTable.create;
 import static java.util.stream.Collectors.toList;
 
-public class AbstractDatatableElementTransformerDefinition extends AbstractGlueDefinition {
+class AbstractDatatableElementTransformerDefinition extends AbstractGlueDefinition {
     private final String[] emptyPatterns;
 
     AbstractDatatableElementTransformerDefinition(Object body, StackTraceElement location, String[] emptyPatterns) {

--- a/java8/src/test/java/io/cucumber/java8/LambdaStepDefinitions.java
+++ b/java8/src/test/java/io/cucumber/java8/LambdaStepDefinitions.java
@@ -8,6 +8,7 @@ import java.util.Optional;
 
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertNotSame;
+import static org.junit.jupiter.api.Assertions.assertNull;
 import static org.junit.jupiter.api.Assertions.assertSame;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
@@ -58,6 +59,7 @@ public class LambdaStepDefinitions implements io.cucumber.java8.En {
             List<Person> people = peopleTable.asList(Person.class);
             assertEquals("Hellesøy", people.get(0).last);
             assertEquals("", people.get(1).last);
+            assertNull(people.get(3).last);
         });
 
         Integer alreadyHadThisManyCukes = 1;

--- a/java8/src/test/resources/io/cucumber/java8/lambda-step-definitions.feature
+++ b/java8/src/test/resources/io/cucumber/java8/lambda-step-definitions.feature
@@ -21,6 +21,7 @@ Feature: Java8
       | Aslak  | Hellesøy |
       | Plato  | [blank]  |
       | Donald | Duck     |
+      | Toto   |          |
     And A string generic that is not a data table
 
   Scenario: using method references

--- a/pom.xml
+++ b/pom.xml
@@ -43,7 +43,7 @@
         <apiguardian-api.version>1.1.0</apiguardian-api.version>
         <!--Dependencies -->
         <cucumber-expressions.version>8.3.1</cucumber-expressions.version>
-        <datatable.version>3.2.0</datatable.version>
+        <datatable.version>3.2.1</datatable.version>
         <tag-expressions.version>2.0.4</tag-expressions.version>
         <messages.version>9.0.3</messages.version>
         <gherkin-vintage.version>5.2.0</gherkin-vintage.version>
@@ -382,6 +382,17 @@
                                 </versionIncreaseAllows>
                             </revapi.semver.ignore>
                             <revapi.ignore>
+                                <item>
+                                    <code>java.class.visibilityReduced</code>
+                                    <old>class io.cucumber.java.AbstractDatatableElementTransformerDefinition</old>
+                                    <justification>Internal API</justification>
+                                </item>
+                                <item>
+                                    <code>java.class.visibilityReduced</code>
+                                    <old>class io.cucumber.java8.AbstractDatatableElementTransformerDefinition</old>
+                                    <justification>Internal API</justification>
+                                </item>
+
                                 <!-- Gherkin parser related -->
                                 <item>
                                     <code>java.class.externalClassExposedInAPI</code>

